File: src/cotmatrix.cpp
#include "cotmatrix.hpp"

Eigen::SparseMatrix<double>
trimesh_cotmatrix(
    Eigen::Ref<const compas::RowMatrixXd> V,
    Eigen::Ref<const compas::RowMatrixXi> F
) {
    Eigen::SparseMatrix<double> L;
    igl::cotmatrix(V, F, L);
    return L;
}

compas::RowMatrixXd
trimesh_cotmatrix_entries(
    Eigen::Ref<const compas::RowMatrixXd> V,
    Eigen::Ref<const compas::RowMatrixXi> F
) {
    Eigen::MatrixXd C;
    igl::cotmatrix_entries(V, F, C);
    compas::RowMatrixXd C_row = C;
    return C_row;
}

NB_MODULE(_cotmatrix, m) {

    m.def(
        "trimesh_cotmatrix",
        &trimesh_cotmatrix,
        "Compute the cotangent Laplacian matrix for a triangle mesh.",
        "V"_a, "F"_a
    );

    m.def(
        "trimesh_cotmatrix_entries",
        &trimesh_cotmatrix_entries,
        "Compute cotangent values for each edge in each triangle.",
        "V"_a, "F"_a
    );
}

File: src/compas_libigl/cotmatrix.py
import numpy as np
from compas.plugins import plugin

from compas_libigl import _cotmatrix


@plugin(category="trimesh")
def trimesh_cotmatrix(M):
    """Compute the cotangent Laplacian matrix of a triangle mesh.

    Parameters
    ----------
    M : tuple[list[list[float]], list[list[int]]]
        A mesh represented by a tuple of (vertices, faces)
        where vertices are 3D points and faces are triangles

    Returns
    -------
    scipy.sparse.csc_matrix
        The cotangent Laplacian matrix in sparse format.
    """
    V, F = M
    V = np.asarray(V, dtype=np.float64)
    F = np.asarray(F, dtype=np.int32)
    return _cotmatrix.trimesh_cotmatrix(V, F)


@plugin(category="trimesh")
def trimesh_cotmatrix_entries(M):
    """Compute cotangent values for each edge in each triangle.

    Parameters
    ----------
    M : tuple[list[list[float]], list[list[int]]]
        A mesh represented by a tuple of (vertices, faces)
        where vertices are 3D points and faces are triangles

    Returns
    -------
    numpy.ndarray
        A matrix of shape (F, 3) containing cotangent values per edge.
        For each face, contains cotan of angles opposite to edges (1,2), (2,0), (0,1).
    """
    V, F = M
    V = np.asarray(V, dtype=np.float64)
    F = np.asarray(F, dtype=np.int32)
    return _cotmatrix.trimesh_cotmatrix_entries(V, F)
